File: src/ui/widgets/CommsWidget.js
// src/ui/widgets/CommsWidget.js

export class CommsWidget {
  constructor({ id = "comms-widget", ctx, onMessageClick }) {
    this.id = id;
    this.ctx = ctx;
    this.onMessageClick = onMessageClick;
    this.messages = [];
    this.enabled = true;
    this.filterShipId = null;

    this.el = document.createElement("div");
    this.el.className = "comms-widget";
    
    Object.assign(this.el.style, {
      display: "flex",
      alignItems: "center",
      gap: "8px",
      pointerEvents: "auto",
      padding: "4px 12px",
      background: "rgba(10, 15, 25, 0.6)",
      borderRadius: "6px",
      border: "1px solid rgba(100, 150, 255, 0.2)",
      fontFamily: "system-ui, -apple-system, sans-serif",
      fontSize: "12px",
      color: "#e0e0e0",
      maxWidth: "400px",
      cursor: "pointer",
      transition: "all 0.2s ease",
    });

    this._injectStyles();
  }

  _injectStyles() {
    if (document.getElementById("comms-widget-style")) return;
    const style = document.createElement("style");
    style.id = "comms-widget-style";
    style.textContent = `
      .comms-entry {
        display: flex;
        align-items: center;
        gap: 8px;
      }
      .comms-sender {
        font-weight: 600;
        font-size: 11px;
        opacity: 0.9;
        white-space: nowrap;
      }
      .comms-text {
        line-height: 1.3;
        font-size: 11px;
        flex: 1;
      }
      .comms-widget:hover {
        background: rgba(20, 30, 50, 0.8);
        border-color: rgba(100, 150, 255, 0.4);
      }
    `;
    document.head.appendChild(style);
  }

  mount(parent, props) {
    parent.appendChild(this.el);
  }

  setVisible(visible) {
    this.enabled = visible;
    this.el.style.display = visible ? "flex" : "none";
  }

  update(game, scene, dt) {}
  render(game, scene, rect) {}

  addMessage(msg) {
    const entry = {
      id: `${Date.now()}_${Math.random()}`,
      type: msg.type || 'system',
      shipId: msg.shipId || null,
      shipName: msg.shipName || null,
      text: msg.text || '',
      action: msg.action || null,
      timestamp: msg.timestamp || Date.now(),
    };

    this.messages.push(entry);
    
    if (this.messages.length > 50) {
      this.messages.shift();
    }

    this._render();
  }

  removeShipMessages(shipId) {
    this.messages = this.messages.filter(m => m.shipId !== shipId);
    this._render();
  }

  showShipHistory(shipId) {
    this.filterShipId = shipId;
    this._render();
  }
  markAsRead(shipId) {
    // В текущей реализации показывается только последнее сообщение,
    // поэтому этот метод оставлен как заглушка для совместимости.
  }

  clearFilter() {
    this.filterShipId = null;
    this._render();
  }

  _render() {
    if (!this.el) return;

    this.el.innerHTML = "";

    // Показываем только одно последнее сообщение
    const messagesToShow = this.filterShipId
      ? this.messages.filter(m => m.shipId === this.filterShipId)
      : this.messages;

    if (messagesToShow.length === 0) {
      const empty = document.createElement("div");
      empty.style.cssText = "opacity: 0.5; font-size: 11px;";
      empty.textContent = "Нет сообщений";
      this.el.appendChild(empty);
      return;
    }

    const lastMsg = messagesToShow[messagesToShow.length - 1];

    const entry = document.createElement("div");
    entry.className = "comms-entry";

    const senderSpan = document.createElement("span");
    senderSpan.className = "comms-sender";
    senderSpan.textContent = lastMsg.shipName || (lastMsg.type === 'system' ? 'Система' : 'Неизвестно');

    const textSpan = document.createElement("span");
    textSpan.className = "comms-text";
    textSpan.textContent = lastMsg.text;

    entry.appendChild(senderSpan);
    entry.appendChild(textSpan);

    this.el.appendChild(entry);

    // Клик по виджету
    this.el.onclick = () => {
      if (lastMsg.shipId && this.onMessageClick) {
        this.onMessageClick(lastMsg);
      }
    };
  }

  destroy() {
    this.messages = [];
    try { this.el?.remove(); } catch (_) {}
  }
}

File: c_cpp_migration/cpp_project/ui/widgets/CommsWidget.cpp
#ifndef COMMSWIDGET_HPP
#define COMMSWIDGET_HPP

#include <iostream>
#include <string>
#include <vector>
#include <memory>
#include <unordered_map>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <optional>
#include <functional>
#include <random>

namespace lostjump {

class CommsWidget {
public:
    // Constructor
    CommsWidget();
};

} // namespace lostjump

#endif // COMMSWIDGET_HPP

// Implementation
namespace lostjump {

#include <iostream>
#include <string>
#include <vector>
#include <memory>
#include <unordered_map>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <optional>
#include <functional>
#include <random>


function apply(el, styles) { Object.assign(el.style, styles); }

class CommsWidget {
  CommsWidget({ id = "comms-widget", ctx } = {}) {
    this.id = id;
    this.ctx = ctx;
    this.el = nullptr;
    this._last = "";
  }

  mount(parent) {
    const el = document.createElement("div");
    this.el = el;
    parent.appendChild(el);

    apply(el, {
      pointerEvents: "none",
      width: "380px",
      color: "#ecf3ff",
      background: "linear-gradient(180deg, rgba(15,20,34,0.84), rgba(8,11,19,0.84))",
      border: "1px solid rgba(180, 213, 255, 0.18)",
      borderRadius: "16px",
      boxShadow: "0 10px 28px rgba(0,0,0,0.4)",
      backdropFilter: "blur(8px)",
      padding: "12px",
      fontFamily: "Inter, system-ui, -apple-system, Segoe UI, Roboto, Arial, sans-serif",
      fontSize: "12px",
      lineHeight: "1.3",
    });

    el.innerHTML = `<div data-k="content">Подключение к каналу связи…</div>`;
    this.$content = el.querySelector('[data-k="content"]');
  }

  setVisible(v) {
    if (!this.el) return;
    this.el.style.display = v ? "" : "none";
  }

  update() {
    const logs = this.ctx.quest.log value_or([];
    const recent = logs.slice(-3).map([](auto& item){ return (item; }) => item.text).filter([](auto& item){ return Boolean; });
    const lines = recent.size() ? recent : [this.ctx.lastLog || "Ожидание сообщений…"];

    const next =
      `<div style="font-size:10px;letter-spacing:.14em;opacity:.72;margin-bottom:8px;">КОММУНИКАЦИИ</div>` +
      lines.map([](auto& item){ return (line; }) => `<div style="padding:6px 8px;border-radius:8px;background:rgba(255,255,255,.05);margin-top:6px;">${line}</div>`).join("");

    if (next === this._last) return;
    this._last = next;
    this.$content.innerHTML = next;
  }

  destroy() {
    try { this.el.remove(); } catch (_) {}
    this.el = nullptr;
  }
}


} // namespace lostjump
